frontend/app/src/next/learn/hooks/use-lesson.tsx

```diff
@@ -0,0 +1,172 @@
+import {
+  useMemo,
+  useState,
+  useRef,
+  useEffect,
+  createContext,
+  useContext,
+  useCallback,
+} from 'react';
+import {
+  HighlightStateMap,
+  SupportedLanguage,
+  LessonPlan,
+  PackageManager,
+  HighlightState,
+} from '@/next/learn/components';
+
+interface ExtraWithPackageManager {
+  packageManager: PackageManager;
+}
+
+export interface LessonState<
+  S extends string,
+  E extends ExtraWithPackageManager,
+  C,
+> {
+  language: SupportedLanguage;
+  setLanguage: (language: SupportedLanguage) => void;
+  activeStep: S;
+  setActiveStep: (step: S) => void;
+  extra: Partial<E>;
+  setExtra: (e: Partial<E>) => void;
+  highlights: HighlightStateMap<S>;
+  setHighlights: (state?: HighlightStateMap<S>) => void;
+  commands: C;
+  stepKeys: S[];
+  currentStepIndex: number;
+  codeBlocksRef: React.MutableRefObject<Record<S, HTMLDivElement | null>>;
+  stepCardsRef: React.MutableRefObject<Record<S, HTMLDivElement | null>>;
+  codeKeyFrames: Record<string, Partial<Record<S, HighlightState>>>;
+}
+
+export interface LessonProviderProps<
+  S extends string,
+  E extends ExtraWithPackageManager,
+  C,
+> {
+  children: React.ReactNode;
+  lesson: LessonPlan<S, E, C>;
+}
+
+const LessonContext = createContext<LessonState<any, any, any> | null>(null);
+
+export function LessonProvider<
+  S extends string,
+  E extends ExtraWithPackageManager,
+  C,
+>({ children, lesson }: LessonProviderProps<S, E, C>) {
+  const [language, setLanguage] = useState(lesson.defaultLanguage);
+  const [activeStep, setActiveStepState] = useState<S>();
+  const [extra, setExtra] = useState(lesson.extraDefaults);
+  const [highlights, setHighlightState] = useState<HighlightStateMap<S>>({});
+
+  const codeBlocksRef = useRef<Record<S, HTMLDivElement | null>>(
+    Object.keys(lesson.steps).reduce(
+      (acc, key) => ({
+        ...acc,
+        [key]: null,
+      }),
+      {} as Record<S, HTMLDivElement | null>,
+    ),
+  );
+
+  const stepCardsRef = useRef<Record<S, HTMLDivElement | null>>(
+    Object.keys(lesson.steps).reduce(
+      (acc, key) => ({
+        ...acc,
+        [key]: null,
+      }),
+      {} as Record<S, HTMLDivElement | null>,
+    ),
+  );
+
+  const stepKeys = Object.keys(lesson.steps) as S[];
+  const currentStepIndex = activeStep ? stepKeys.indexOf(activeStep) : -1;
+
+  useEffect(() => {
+    console.log('activeStep', activeStep);
```

Comment:
I'm surprised these log lines aren't caught by the linter?

Reply:
me too